docs/testing.md

```diff
@@ -0,0 +1,136 @@
+
+To run tests against either Postgres or MySQL, use the `postgresql` or `mysql` profile, to select the DB. Be sure
+to add the `default` profile which will trigger seeding the database with some admin users, clients, etc. For example:
+
+    $ ./gradlew '-Dspring.profiles.active=mysql,default test
+
+To run tests from your IDE against a given database, you can (temporarily) annotate the test class:
+
+```java
+
+@ActiveProfiles({"mysql", "default"})
+class MyCustomTests {
+    @Test
+    void foo() {
+        // ...
+    }
+}
+```
+
+## Database-specific tests
+
+Some tests only work on a single type of database, for example `MySqlDbMigrationIntegrationTest`; or do not work on a
+given database, for example `JdbcClientMetadataProvisioningTest.createdByPadsTo36Chars`. You can turn tests on and off
+based on the profile with custom annotations, `@DisabledIfProfile` and `@EnabledIfProfile`, for example:
+
+```java
+// Only run on either mysql or postgresql
+@EnabledIfProfile({"mysql", "postgresql"})
+class RealDbOnlyTests {
+
+}
+
+// or:
+
+class SomeTests {
+
+    // Do not run when there is either mysql or postgresql
+    @DisabledIfProfile({"mysql", "postgres"})
+    void notOnRealDb() {
+
+    }
+
+}
+```
+
+## Test pollution
+
+There might be test pollution when tests are run in parallel, or even between projects. For example, when you run
+
+    $ ./gradlew test
+
+It will run tests in both `cloudfoundry-identity-server` and `cloudfoundry-identity-uaa` projects. Both need a database,
+and both do sometimes clean up the database.
+
+To avoid test pollution, 24 databases are created, and each Gradle "worker" thread gets its own database. A Gradle
+worker has a numeric `id`, and each time a new task is spun up, the idea of the worker picking up the task increases.
+So there are 24 DBs with names `uaa_1`, `uaa_2`, ... created, and usually the worker ID stays below 24 and there are
+enough databases for each test.
+
+However, if the gradle daemon is kept running in the background and is re-used for subsequent tasks, e.g. by doing:
+
+    $ ./gradlew test # first run
+    # do some code changes
+    $ ./gradlew test
+
+You will get new workers with IDs > 24. It is recommended you run your Gradle in no-daemon mode when running tests:
+
+    $ ./gradlew test --no-daemon
+
+It will be slightly slower to start up (a few seconds), but the tests take multiple minutes and so the gain of using
+a daemon is not worth the trouble.
+
+## Timezone issues
+
+The UAA and its DB server _MUST_ have the same timezone, because dates are not uniformly stored in UTC and timezones
+do matter. Specifically for MySQL, there are issues when your local host is ahead of UTC, because:
+
+1. The default containers runs in UTC
+2. When calling `current_timestamp` the value is in UTC
+3. But when calling a prepared statement from JDBC with a Date/Timestamp/time-based the timezone is sent to the server
+
+So, when running e.g. in `Europe/Paris` (CET):
```

Comment:
:-) its ok, but what about Europe/Berlin or Europe/Madrid ? ... I try to omit such country specifics...but CET is good enough to call it Europe

Reply:
Yes, but also:
- `Europe/Lisbon` (Western European Time, WEST) is UTC / UTC+1
- `Europe/Bucharest` (Eastern European Time, EEST) is UTC+2 / UTC+3

And if you truly want to be nitpicky:
-  There's even a `Europe/Moscow` (+3/+5) and `Europe/Samara (+4) 😱 

Promise I'll do `Europe/Berlin` next time!

Reply:
ok, maybe I explained my intension for "So, when running e.g. in CET" not good enough.

CET is enough, we dont need further information like the country or city.
You use UTC also without context like Earth (UTC)

Reply:
Ah yes makes sense, I get it. My bad.

To clarify, when I say "[e.g.](https://simple.wikipedia.org/wiki/Exempli_gratia) in" it means "for example in"

Reply:
I understood for example, but TMI